first_order_model/shrink_util.py

```diff
@@ -0,0 +1,1198 @@
+        if pruners[0] == 'first':
+            for prune_indices in reverse_sort(pruners[1:]):
+                if node.type == 'conv' and len(node.after) != 0:
+                    # Prune the outupts of the convolutional layer
+                    f_set(node.value.weight.detach(), node.value.weight, prune_indices)
+                    if node.value.bias is not None:
+                        f_set(node.value.bias.detach(), node.value.bias, prune_indices)
+                else:
+                    print("Should not be shrinking a batchnorm")
+                # Hacky solution to modify group count for dethwise convolutions
+                if node.value.groups != 1:
+                    node.value.groups = node.value.weight.shape[0]
+
+        else: # Delete from its inputs
+            for prune_indices in reverse_sort(pruners):
+
+                # Prune inputs of a convolutional layer
+                # We ignore depthwise layers since they take 1 input, their groups are just what changes.
+                if node.type == 'conv' and node.value.groups == 1:
+                    nvd = node.value.weight.detach()
+                    nvd = torch.cat(
+                        [nvd[:, :prune_indices[0]], nvd[:, prune_indices[1]:]],
+                        dim=1)
+                    node.value.weight.set_(nvd.clone().detach().contiguous())
+                
+                # Prune the inputs (and by definition outputs) of a batchnorm layer
+                if node.type == 'bn':
+
+                    f_set(node.value.weight.detach(), node.value.weight,
+                          prune_indices)
+                    f_set(node.value.bias.detach(), node.value.bias,
+                          prune_indices)
+                    f_set(node.value.running_mean.detach(),
+                          node.value.running_mean, prune_indices)
+                    f_set(node.value.running_var.detach(),
+                          node.value.running_var, prune_indices)
+
+    return model
+
+
+def get_metrics_loss(metrics_dataloader, lr_size, generator_full,
+                     generator_type):
+    """
+    Computes total loss when running generator_full on metrics_dataloader's all inputs
+    """
+    total_loss = 0
+    with torch.no_grad():
+        for y in metrics_dataloader:
+            y['driving_lr'] = F.interpolate(y['driving'], lr_size)
+            
+            # Inputs need to be moved manually to gpu
+            move_to_gpu(y)
+            losses_generator, metrics_generated = generator_full(
+                y, generator_type)
+            loss_values = [val.mean() for val in losses_generator.values()]
+            loss = sum(loss_values)
+            total_loss += loss.item()
+
+    return total_loss
+
+def get_following_layers_skip_depthwise(following_layers, layer_graph, x):
+    """
+    Specialized version of follo that skips over depthwise convolved layers. Only used when sorting the outputs of a layer for importance.
```

Comment:
typo: follow

Reply:
Technically yes, but they are two different cases

Reply:
that's fine please minimize code copying, so use them as closely as possible

Reply:
Removed duplicated cases